runtime/compiler/control/CompilationThread.cpp

```diff
@@ -5982,7 +5982,15 @@ void *TR::CompilationInfo::compileOnSeparateThread(J9VMThread * vmThread, TR::Il
 
    if (TR::Options::sharedClassCache() && !TR::Options::getAOTCmdLineOptions()->getOption(TR_NoLoadAOT) && details.isOrdinaryMethod())
       {
-      if (!isJNINative(method) && !oldStartPC)
+      bool firstTimeCompile = !oldStartPC;
+
+      // Under AOT, FSD and Sync Compilation, a hang is possible
+      // (see eclipse/openj9#11879 for more details).
+      // Therefore, use the isCompiled(method) query.
+      if (TR::Options::getCmdLineOptions()->getOption(TR_FullSpeedDebug) && !async)
+         firstTimeCompile = !isCompiled(method);
```

Comment:
I am thinking that we should always use `isCompiled(method)` rather than `!oldStartPC`. The code looks cleaner that way.

Reply:
Updated in https://github.com/eclipse/openj9/pull/11879/commits/6133b4b2ea4485d876a0e4e39d0b1d0ae53fb18e